src/pgen/cluster/agn_feedback.cpp

```diff
@@ -0,0 +1,230 @@
+  const auto &cons_pack = md->PackVariables(std::vector<std::string>{"cons"});
+  IndexRange ib = md->GetBlockData(0)->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = md->GetBlockData(0)->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = md->GetBlockData(0)->GetBoundsK(IndexDomain::interior);
+
+  // Thermal quantities
+  const Real thermal_power = power * thermal_fraction_;
+  const Real thermal_scaling_factor = 1 / (4. / 3. * M_PI * pow(thermal_radius_, 3));
+  const Real thermal_feedback =
+      thermal_power * thermal_scaling_factor * beta_dt; // energy/volume
+  const Real thermal_density_rate =
+      mass_rate * thermal_scaling_factor * thermal_fraction_;
+  const Real thermal_radius2 = thermal_radius_ * thermal_radius_;
+
+  // Kinetic Jet Quantities
+  const Real kinetic_power = power * kinetic_fraction_;
+  const Real kinetic_scaling_factor =
+      1 / (2 * kinetic_jet_height_ * M_PI * pow(kinetic_jet_radius_, 2));
+  // Matches 1/2.*jet_density*jet_velocity*jet_velocity*beta_dt;
+  const Real kinetic_feedback =
+      kinetic_power * kinetic_scaling_factor * beta_dt; // energy/volume
+
+  // Note that new mass is injected to create the kinetic power, separate from the
+  // existing gas
+  const Real kinetic_jet_total_mass = mass_rate * kinetic_fraction_;
+  const Real kinetic_jet_density_rate =
+      kinetic_jet_total_mass * kinetic_scaling_factor * beta_dt;
+  const Real kinetic_jet_velocity =
+      std::sqrt(2. * kinetic_power / kinetic_jet_total_mass);
+  const Real kinetic_jet_momentum_rate = kinetic_jet_density_rate * kinetic_jet_velocity;
+
+  const Real kinetic_jet_radius = kinetic_jet_radius_;
+  const Real kinetic_jet_height = kinetic_jet_height_;
+
+  const parthenon::Real time = tm.time;
+
+  const auto &jet_coords_factory =
+      hydro_pkg->Param<JetCoordsFactory>("jet_coords_factory");
+  const JetCoords jet_coords = jet_coords_factory.CreateJetCoords(time);
+
+  // Constant volumetric heating
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "HydroAGNFeedback::FeedbackSrcTerm",
+      parthenon::DevExecSpace(), 0, cons_pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s,
+      ib.e, KOKKOS_LAMBDA(const int &b, const int &k, const int &j, const int &i) {
+        auto &cons = cons_pack(b);
+        auto &prim = prim_pack(b);
+        const auto &coords = cons_pack.GetCoords(b);
+
+        const Real x = coords.Xc<1>(i);
+        const Real y = coords.Xc<2>(j);
+        const Real z = coords.Xc<3>(k);
+
+        // Thermal Feedback
+        if (thermal_power > 0) {
+          const Real r2 = x * x + y * y + z * z;
+          // Determine if point is in sphere r<=thermal_radius
+          if (r2 <= thermal_radius2) {
+            // Add density at constant velocity and temperature
+            // HACK Disabling injection of mass with thermal feedback since it
+            // leads to small timesteps
+            //AddDensityToConsAtFixedVelTemp(thermal_density_rate, cons, prim, eos, k, j,i);
```

Comment:
Probably? There's a question about strict accounting for mass removed during AGN triggering. Should we inject that mass back in with feedback minus the mass converted to energy? Maybe take some of that mass and add it to the SMBH mass?

For the commit here, I've been adding that mass back proportionally to the feedback mechanisms (i.e., 1/3 the energy goes into kinetic energy, 1/3 the mass gets injected with the kinetic jet). I hacked off the thermal mass injection because there was a bug, the thermal density injection was missing `dt`. 

I've fixed that bug in a commit I was making and testing last week. I also modified the mass injection to account for the mass-to-energy conversion (which is negligible) but more importantly for the proportion of mass injection to be different  from energy injection (i.e. 1/3 energy into the kinetic jet, 1/8 the mass into the kinetic jet). This turns out to be necessary to change the jet velocity due by disconnecting the kinetic energy injected and mass injected.

On the topic of injecting mass with thermal feedback, it might avoid some low density scenarios. Dumping thermal energy near the AGN it tends to blow out the gas. Maybe by injecting some of the AGN triggering gas we can avoid that.